From 96506b0d2b42ae6943e3009120792f71b13d3315 Mon Sep 17 00:00:00 2001
From: Christian Bruel <christian.bruel@foss.st.com>
Date: Fri, 5 Jan 2024 14:54:46 +0100
Subject: [PATCH 0777/1141] PCI: stm32: add PCIe RC pm runtime support

Enable dynamic runtime PM domain support to the RC PCIe driver.
Resumed in probe and released in pcie_remove

Signed-off-by: Christian Bruel <christian.bruel@foss.st.com>
Change-Id: Ie08b60d2a9b646b86ab64ea7e7db1be4f9571461
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/349510
Reviewed-by: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
Domain-Review: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/pci/controller/dwc/pcie-stm32.c | 20 +++++++++++++++++++-
 1 file changed, 19 insertions(+), 1 deletion(-)

--- a/drivers/pci/controller/dwc/pcie-stm32.c
+++ b/drivers/pci/controller/dwc/pcie-stm32.c
@@ -14,6 +14,7 @@
 #include <linux/of_gpio.h>
 #include <linux/phy/phy.h>
 #include <linux/msi.h>
+#include <linux/pm_runtime.h>
 #include <linux/regmap.h>
 #include <linux/reset.h>
 #include "pcie-designware.h"
@@ -381,12 +382,27 @@ static int stm32_pcie_probe(struct platf
 
 	platform_set_drvdata(pdev, stm32_pcie);
 
+	ret = devm_pm_runtime_enable(dev);
+	if (ret < 0) {
+		dev_err(dev, "Failed to enable pm runtime %d\n", ret);
+		goto phy_disable;
+	}
+
+	ret = pm_runtime_resume_and_get(dev);
+	if (ret < 0) {
+		dev_err(dev, "pm runtime resume failed: %d\n", ret);
+		goto phy_disable;
+	}
+
 	ret = stm32_add_pcie_port(stm32_pcie, pdev);
 	if (ret)
-		goto phy_disable;
+		goto sync_disable;
 
 	return 0;
 
+sync_disable:
+	pm_runtime_put_sync(dev);
+
 phy_disable:
 	phy_exit(stm32_pcie->phy);
 
@@ -402,6 +418,8 @@ static int stm32_pcie_remove(struct plat
 	clk_disable_unprepare(stm32_pcie->clk);
 	pcie_port_irqs_hook = NULL;
 
+	pm_runtime_put_sync(&pdev->dev);
+
 	phy_exit(stm32_pcie->phy);
 
 	return 0;
